File: 05-sandbox-execution/golang/agent.go
package main

import (
	"context"
	"fmt"
	"os"
	"sync"

	"byoo/xclaw/providers"
)

const defaultMaxIterations = 10

// Agent manages per-session conversation history and drives the agent loop.
type Agent struct {
	mu            sync.Mutex
	sessions      map[string][]providers.Message
	providerChain []string
	maxIterations int
}

func newAgent(providerChain []string, maxIterations int) *Agent {
	if maxIterations <= 0 {
		maxIterations = defaultMaxIterations
	}
	return &Agent{
		sessions:      make(map[string][]providers.Message),
		providerChain: providerChain,
		maxIterations: maxIterations,
	}
}

// handle processes one user message. onDelta is called for each streamed token
// once we confirm the response is a plain text reply (not a tool call).
func (a *Agent) handle(msg ACPMessage, onDelta func(string)) (string, error) {
	a.mu.Lock()
	if _, ok := a.sessions[msg.SessionID]; !ok {
		a.sessions[msg.SessionID] = []providers.Message{
			{Role: "system", Content: buildSystemPrompt()},
		}
	}
	messages := make([]providers.Message, len(a.sessions[msg.SessionID]))
	copy(messages, a.sessions[msg.SessionID])
	a.mu.Unlock()

	messages = append(messages, providers.Message{Role: "user", Content: msg.Content})

	for iterations := 0; iterations < a.maxIterations; iterations++ {
		// Buffer tokens — only flush to onDelta when confirmed as a text reply.
		var buffer []string
		reply, err := providers.StreamWithFallback(
			context.Background(), messages, a.providerChain, assemble,
			func(token string) { buffer = append(buffer, token) },
		)
		if err != nil {
			return "", err
		}
		messages = append(messages, providers.Message{Role: "assistant", Content: reply})

		toolCall := extractJSON(reply)
		action, isToolCall := "", false
		if toolCall != nil {
			if a, ok := toolCall["action"].(string); ok && a != "" {
				action = a
				isToolCall = true
			}
		}

		if !isToolCall {
			// Plain text reply — flush buffered tokens to the channel.
			for _, t := range buffer {
				onDelta(t)
			}
			a.mu.Lock()
			a.sessions[msg.SessionID] = messages
			a.mu.Unlock()
			return reply, nil
		}

		tool, found := toolRegistry[action]
		if !found {
			errMsg := fmt.Sprintf(`error: unknown tool "%s". Available: %s`, action, availableTools())
			messages = append(messages, providers.Message{Role: "user", Content: errMsg})
			continue
		}

		params := make(map[string]string)
		for k, v := range toolCall {
			if k == "action" {
				continue
			}
			if s, ok := v.(string); ok {
				params[k] = s
			}
		}

		fmt.Fprintf(os.Stderr, "[agent] [%s] uses [%s]: %v\n", msg.SessionID, action, params)
		output, execErr := tool.Execute(msg.SessionID, params)
		if execErr != nil {
			fmt.Fprintf(os.Stderr, "[agent] tool error: %v\n", execErr)
			messages = append(messages, providers.Message{Role: "user", Content: "tool error:\n" + execErr.Error()})
		} else {
			messages = append(messages, providers.Message{Role: "user", Content: "tool output:\n" + output})
		}
	}

	notice := fmt.Sprintf("[xclaw] reached max iterations (%d), stopping", a.maxIterations)
	a.mu.Lock()
	a.sessions[msg.SessionID] = messages
	a.mu.Unlock()
	return notice, nil
}

File: 05-sandbox-execution/nodejs/src/agent.ts
import { streamWithFallback } from './providers/registry.ts';
import { toolRegistry, buildToolsPrompt, extractJSON } from './tools.ts';
import type { Message } from './providers/types.ts';
import type { ACPMessage } from './gateway/types.ts';
import { log } from './logger.ts';

export class Agent {
  private sessions = new Map<string, Message[]>();
  private providerChain: string[];
  private maxIterations: number;

  constructor(providerChain: string[], maxIterations: number) {
    this.providerChain = providerChain;
    this.maxIterations = maxIterations;
  }

  // Process one user message; call onDelta for each streamed token.
  async handle(msg: ACPMessage, onDelta: (token: string) => void): Promise<string> {
    if (!this.sessions.has(msg.sessionId)) {
      this.sessions.set(msg.sessionId, [{ role: 'system', content: buildSystemPrompt() }]);
    }
    const messages = this.sessions.get(msg.sessionId)!;
    messages.push({ role: 'user', content: msg.content });

    let iterations = 0;
    while (true) {
      if (++iterations > this.maxIterations) {
        const notice = `[xclaw] reached max iterations (${this.maxIterations}), stopping`;
        log(notice);
        return notice;
      }

      const buffer: string[] = [];
      const reply = await streamWithFallback(messages, this.providerChain, (token) => {
        buffer.push(token);
      });
      messages.push({ role: 'assistant', content: reply });

      const toolCall = extractJSON(reply);
      if (toolCall && typeof toolCall['action'] === 'string') {
        const tool = toolRegistry.get(toolCall['action']);
        if (tool) {
          const { action, ...rawParams } = toolCall as Record<string, string>;
          const params: Record<string, string> = {};
          for (const [k, v] of Object.entries(rawParams)) {
            if (typeof v === 'string') params[k] = v;
          }
          log(`[agent] [${msg.sessionId}] uses [${action}]: ${JSON.stringify(params)}`);
          try {
            const output = await tool.execute(msg.sessionId, params);
            log(output);
            messages.push({ role: 'user', content: `tool output:\n${output}` });
          } catch (err: any) {
            const errMsg = err.stderr ?? err.message;
            console.error(`[agent] tool error: ${errMsg}`);
            messages.push({ role: 'user', content: `tool error:\n${errMsg}` });
          }
        } else {
          const available = [...toolRegistry.keys()].join(', ');
          messages.push({ role: 'user', content: `error: unknown tool "${toolCall['action']}". Available: ${available}` });
        }
      } else {
        for (const token of buffer) onDelta(token);
        return reply;
      }
    }
  }
}

function buildSystemPrompt(): string {
  return `You are an AI assistant named xclaw.

To use a tool, output a JSON object (bare or in a markdown code block):
{"action": "<tool_name>", "<param1>": "<value1>", ...}

To answer directly, output plain text — do NOT use JSON.

Available tools:
${buildToolsPrompt()}`;
}
